### PR TITLE
[Jobs] Add option to specify `max_restarts_on_errors`

### PR DESCRIPTION
**Pull request by Michaelvll**
Allow max retry on failure for managed jobs.

## Interface

```
resources:
    job_recovery:
        max_restarts_on_errors: 3

setup: |
    echo setup
    exit 1
```

## Behavior

When `max_restarts_on_errors` is specified, any user program failure will be retried until reaching the max value. The preemption and cluster failure is not counted in the retry.

During restart, the whole cluster is terminated and the job will be started on a new cluster re-provisioned by SkyPilot.

TODO:
- [x] Add unittests for recovery strategy default value setting

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] `sky jobs launch -n test test.yaml --cloud aws --cpus 2+`
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [x] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`
  - [x] Backward compatibility for existing controller with old jobs running.

### CODE REVIEW

File: docs/source/examples/managed-jobs.rst
Comment (concretevitamin) [inline docs/source/examples/managed-jobs.rst:293]:
```suggestion
retry the job on its own failure, e.g., when a training job crashes due to a Nvidia driver issue or NCCL timeouts. To specify this, you
```

File: docs/source/reference/yaml-spec.rst
Comment (concretevitamin) [inline docs/source/reference/yaml-spec.rst:113]: Just looking at naming, it's slightly confusing because 'retry' made me think about retry as in autofailover.

How about we call it `max_restarts / max_restarts_on_failure`?

-----

"Or use a dict to additionally specify maximum job restarts:"

**Review by concretevitamin — COMMENTED** [review]
Quick look on field name.

File: docs/source/reference/yaml-spec.rst
Comment (Michaelvll) [reply docs/source/reference/yaml-spec.rst:113]: Good point! Updated. Thanks

File: sky/resources.py
Comment (concretevitamin) [inline sky/resources.py:166]: Should we update type hint above?

File: docs/source/examples/managed-jobs.rst
Comment (concretevitamin) [inline docs/source/examples/managed-jobs.rst:290]: How about keep this simple in 1 section (along with L304)? Maybe "Job Restarts on User Code Failure"?

File: docs/source/examples/managed-jobs.rst
Comment (concretevitamin) [inline docs/source/examples/managed-jobs.rst:294]:
```suggestion
By default, SkyPilot will try to recover a job when its underlying cluster is preempted or failed. Any user code failures (non-zero exit codes) are not auto-recovered.

In some cases, you may want a job to automatically restart on its own failures, e.g., when a training job crashes due to a Nvidia driver issue or NCCL timeouts. To specify this, you
can further set :code:`max_retry_on_failure` in :code:`resources.job_recovery` in the job YAML file.
```

File: docs/source/examples/managed-jobs.rst
Comment (concretevitamin) [inline docs/source/examples/managed-jobs.rst:301]: Update name?

"Restart up to this many times on user code failures."

**Review by concretevitamin — COMMENTED** [review]
Quick look on interface, LGTM. Maybe @romilbhardwaj should review the impl?

File: docs/source/examples/managed-jobs.rst
Comment (romilbhardwaj) [inline docs/source/examples/managed-jobs.rst:300]: I'm slightly confused by the naming. Just looking at `max_restarts_on_failure` makes me think it also applies to spot instance failure (does it apply? or are spot instance failure max restarts infinite?).

Perhaps a clearer way would be to have two distinct fields like so:
```
  resources:
    accelerators: A100:8
    job_recovery:
      restart_on_job_errors: bool
      max_restarts_on_failure: int, defaults to infinite
```

wdyt? cc @concretevitamin

File: sky/jobs/recovery_strategy.py
Comment (romilbhardwaj) [inline sky/jobs/recovery_strategy.py:384]: The name of this method is a bit misleading since this does not "trigger" a retry, rather just records it. Maybe rename to `log_retry_on_failure` or `record_retry_on_failure`?

Docstr could also be updated to something like:
`"""Records a retry event after a job failure and returns if more retries should be attempted."""`

File: sky/jobs/utils.py
Comment (romilbhardwaj) [inline sky/jobs/utils.py:395]: Is the print here and L417 intentional?

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @Michaelvll! Haven't tried yet, but code mostly looks ok to me. Sending some early comments first.

File: docs/source/examples/managed-jobs.rst
Comment (Michaelvll) [reply docs/source/examples/managed-jobs.rst:299]: How about `max_restarts_on_user_failure`? I found having two fields a bit verbose. Wdyt?

File: sky/jobs/recovery_strategy.py
Comment (Michaelvll) [reply sky/jobs/recovery_strategy.py:384]: Good point! I renamed it to `should_restart_on_failure` with the docstr updated. Wdyt?

File: sky/jobs/utils.py
Comment (Michaelvll) [reply sky/jobs/utils.py:395]: Yes, added a comment. Thanks!

File: docs/source/examples/managed-jobs.rst
Comment (romilbhardwaj) [reply docs/source/examples/managed-jobs.rst:300]: Hmm I think two fields is clearer and clarifies the intent. Not feeling too strongly though. 

If we do decide on having a single field, how about `max_restarts_on_job_errors`?

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @Michaelvll!

File: sky/jobs/recovery_strategy.py
Comment (Michaelvll) [reply sky/jobs/recovery_strategy.py:390]: Ahh, good catch! I was counting the first execution of the job as one of it, but you are right, it is more clear to be actual restarts

File: docs/source/examples/managed-jobs.rst
Comment (Michaelvll) [reply docs/source/examples/managed-jobs.rst:300]: Discussed offline, we will use `max_restarts_on_errors`

**Review by romilbhardwaj — APPROVED** [review]
This is awesome, thanks @Michaelvll! Should be good to go after [docs build is fixed](https://github.com/skypilot-org/skypilot/actions/runs/11564322702/job/32189299480?pr=4169).

File: sky/jobs/utils.py
Comment (andylizf) [inline sky/jobs/utils.py:408]: Seems the retry logic is added in the wrong branch. It's currently in the `else` branch of `if task_id < num_tasks - 1 and follow`, which means it only triggers when we want to terminate. The retry check should be in the outer `else` branch where we handle cluster failures.

**Pull request merged**
